pkg/sync/grpc/grpc_sync.go

```diff
@@ -0,0 +1,126 @@
+
+	"github.com/open-feature/flagd/pkg/logger"
+	"github.com/open-feature/flagd/pkg/sync"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+// Prefix for GRPC URL inputs. GRPC does not define a prefix through standard. This prefix helps to differentiate
+// remote URLs for REST APIs (i.e - HTTP) from GRPC endpoints.
+const Prefix = "grpc://"
+
+type Sync struct {
+	Target string
+	Key    string
+	Logger *logger.Logger
+}
+
+func (g *Sync) Sync(ctx context.Context, dataSync chan<- sync.DataSync) error {
+	dial, err := grpc.Dial(g.Target, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		g.Logger.Error(fmt.Sprintf("Error establishing connection: %s", err.Error()))
+		return err
+	}
+
+	return g.streamListener(ctx, dial, dataSync)
+}
+
+// streamListener performs the grpc listening on provided client connection and push updates through dataSync channel
+func (g *Sync) streamListener(ctx context.Context, dial *grpc.ClientConn, dataSync chan<- sync.DataSync) error {
+	group, localContext := errgroup.WithContext(ctx)
+
+	group.Go(func() error {
+		serviceClient := syncv1grpc.NewFlagServiceClient(dial)
+
+		syncClient, err := serviceClient.SyncFlags(context.Background(), &v1.SyncFlagsRequest{Key: g.Key})
+		if err != nil {
+			g.Logger.Error(fmt.Sprintf("Error calling streaming operation: %s", err.Error()))
+			return err
+		}
+
+		return g.handleFlagSync(syncClient, dataSync)
+	})
+
+	<-localContext.Done()
+
+	err := group.Wait()
+	if err == io.EOF {
+		g.Logger.Info("Stream closed by the server")
+		return err
+	}
+
+	return err
+}
+
+func (g *Sync) handleFlagSync(stream syncv1grpc.FlagService_SyncFlagsClient, dataSync chan<- sync.DataSync) error {
+	for {
+		data, err := stream.Recv()
+		if err != nil {
+			g.Logger.Warn(fmt.Sprintf("Error with stream response: %s", err.Error()))
+			return err
+		}
+
+		switch data.State {
+		case v1.SyncState_SYNC_STATE_ALL:
+			dataSync <- sync.DataSync{
+				FlagData: data.Flags,
+				Source:   g.Target,
+				Type:     sync.ALL,
+			}
+
+			g.Logger.Debug("received full configuration payload")
+		case v1.SyncState_SYNC_STATE_ADD:
+			dataSync <- sync.DataSync{
+				FlagData: data.Flags,
+				Source:   g.Target,
+				Type:     sync.ADD,
+			}
+
+			g.Logger.Debug("received an add payload")
+		case v1.SyncState_SYNC_STATE_UPDATE:
+			dataSync <- sync.DataSync{
+				FlagData: data.Flags,
+				Source:   g.Target,
+				Type:     sync.UPDATE,
+			}
+
+			g.Logger.Debug("received an update payload")
+		case v1.SyncState_SYNC_STATE_DELETE:
+			dataSync <- sync.DataSync{
+				FlagData: data.Flags,
+				Source:   g.Target,
+				Type:     sync.DELETE,
+			}
+
+			g.Logger.Debug("received a delete payload")
+		case v1.SyncState_SYNC_STATE_PING:
+			g.Logger.Debug("received server ping")
+		default:
+			g.Logger.Warn(fmt.Sprintf("receivied unknown state: %s", data.State.String()))
```

Comment:
Yes, this was a test case to check the state handling. I should have actually used `42` 😉 

I am also now 100% sure about the log level here. I opted for warning as this status is unknow and could happen because of an incorrect server impl / or any other unknown. IMO it was better to log a warn than a debug (compared for example to a server ping)

I am open for suggestions here :)